File: src/main/java/com/hedvig/botService/serviceIntegration/slackService/SlackService.java
package com.hedvig.botService.serviceIntegration.slackService;

import com.hedvig.botService.serviceIntegration.slackService.dto.SlackData;
import com.hedvig.botService.services.events.OnboardingCallForQuoteEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Profile;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
@Profile("slack")
public class SlackService {
  private final SlackServiceClient slackServiceClient;
  private static Logger logger = LoggerFactory.getLogger(SlackService.class);

  @Autowired
  SlackService(SlackServiceClient slackServiceClient) {
    this.slackServiceClient = slackServiceClient;
  }

  @EventListener
  public void on(OnboardingCallForQuoteEvent event) {
    String SLACK_CHANNEL = "#call_member_for_quote";

    final String message = String.format(
      "Potential member %s - %s %s tried to sign-up, give them a call on %s",
      event.getMemberId(), event.getFirstName(), event.getLastName() != null ? event.getLastName() : "", event.getPhoneNumber());

    try {
      SlackData slackData = new SlackData();
      slackData.setChannel(SLACK_CHANNEL);
      slackData.setText(message);

      slackServiceClient.sendNotification(slackData);
    } catch(Exception e) {
      logger.error("Cannot send notification to slack channel {} with message {}", SLACK_CHANNEL, message);
    }
  }
}

File: src/main/java/com/hedvig/botService/serviceIntegration/slackService/SlackService.kt
package com.hedvig.botService.serviceIntegration.slackService

import com.hedvig.botService.serviceIntegration.slackService.dto.SlackData
import com.hedvig.botService.services.events.HouseUnderwritingLimitCallMeExceedsEvent
import com.hedvig.botService.services.events.OnboardingCallForQuoteEvent
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Profile
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

@Component
@Profile("slack")
class SlackService @Autowired
internal constructor(
    private val slackServiceClient: SlackServiceClient,
    private val slackUnderwritingServiceClient: SlackUnderwritingServiceClient
) {

    @EventListener
    fun on(event: OnboardingCallForQuoteEvent) {
        val slackChannel = "#call_member_for_quote"

        val message =
            "Potential member ${event.memberId} - ${event.firstName} ${event.lastName} tried to sign-up, give them a call on ${event.phoneNumber}"

        try {
            val slackData = SlackData(
                slackChannel,
                message
            )

            slackServiceClient.sendNotification(slackData)
        } catch (e: Exception) {
            logger.error("Cannot send notification to slack channel {} with message {}", slackChannel, message)
        }

    }

    @EventListener
    fun on(event: HouseUnderwritingLimitCallMeExceedsEvent) {
        val slackChannel = "#call_member_underwriting_guidelines"

        val message =
            "Member ${event.memberId} - ${event.firstName} ${event.lastName} house exceeds underwriting guidelines" +
              " on reason [${event.reason}], call the member on ${event.phoneNumber}"

        try {
            val slackData = SlackData(
                slackChannel,
                message
            )

            slackUnderwritingServiceClient.sendNotification(slackData)
        } catch (e: Exception) {
            logger.error("Cannot send notification to slack channel {} with message {}", slackChannel, message)
        }

    }

    private val logger = LoggerFactory.getLogger(javaClass)
}
